docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,187 @@
+
+#### Use-case: Clustering a MongoDB database
+
+MongoDB can be clustered with homogeneous peers which perform leader elections. Users should be able to define a MongoDB deployment in which all replicas are automatically clustered.
+
+###### Container lifecycle hooks
+
+TODO: how would the containers be made aware of one another?
+
+###### Post-deployment hooks
+
+Post-deployment hooks could used to execute a command once per logical deployment against each database pod to register cluster members.
+
+#### Use-case: Running Rails migrations
+
+New revisions of a Rails application often contain schema updates which must accompany the new code deployment. Users should be able to specify a hook which performs a Rails migration as part of the application code deployment.
+
+Database migrations are complex and introduce downtime concerns. Here are [some examples](https://blog.rainforestqa.com/2014-06-27-zero-downtime-database-migrations) of zero-downtime Rails migration workflows.
+
+Deployments including database migrations must make special considerations:
+
+1.  Deployed code must be consistent with the currently deployed schema at any point in time.
+2.  Database or table locking must be minimized or eliminated to prevent service outages.
+
+The workflows which are effective at ensuring zero downtime migrations are typically multi-phased. For a user orchestrating a zero downtime migration deployment, it's likely the user needs to verify each deployment step discretely, with the option to abort and rollback after each phase.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks introduce problems with Rails migrations:
+
+1. The migration hook will execute once per pod in the deployment. This is a race and can result in concurrent migration attempts against a shared database. This burdens the hook author with coordinating the hooks or otherwise safeguarding the migration code to ensure it is safe for concurrent execution.
+
+###### Post-deployment hooks
+
+Post-deployment hooks satisfy this use case by providing a means to execute the hook only once per logical deployment.
+
+This approach has a potential caveat: Since post-deployment hooks are executed against an existing container in the deployment, it is implied that the migration would execute in the same process space as a running application container. The migration will thus consume the resources (e.g. CPU, memory) of the application container. Experience with Rails migrations in OpenShift has shown this resource sharing to be problematic with many Rails applications.
+
+This proposal does not provide a solution to account for the resource caveat.
+
+## Hook executable availability
+
+A hook command is executed directly within an existing container in a pod. The executable specified in the hook command must exist in the target container. The executable can be made available in the container in a couple of ways:
```

Comment:
It would be good to break these use cases into what needs to call whom where (rails web container needs to call a script in its own image, using the same config as the pod, to run).


> On Mar 4, 2015, at 5:21 PM, Dan Mace <notifications@github.com> wrote:
> 
> In docs/proposals/post-deployment-hooks.md:
> 
> > +
> > +Container lifecycle hooks introduce problems with Rails migrations:
> > +
> > +1. The migration hook will execute once per pod in the deployment. This is a race and can result in concurrent migration attempts against a shared database. This burdens the hook author with coordinating the hooks or otherwise safeguarding the migration code to ensure it is safe for concurrent execution.
> > +
> > +###### Post-deployment hooks
> > +
> > +Post-deployment hooks satisfy this use case by providing a means to execute the hook only once per logical deployment.
> > +
> > +This approach has a potential caveat: Since post-deployment hooks are executed against an existing container in the deployment, it is implied that the migration would execute in the same process space as a running application container. The migration will thus consume the resources (e.g. CPU, memory) of the application container. Experience with Rails migrations in OpenShift has shown this resource sharing to be problematic with many Rails applications.
> > +
> > +This proposal does not provide a solution to account for the resource caveat.
> > +
> > +## Hook executable availability
> > +
> > +A hook command is executed directly within an existing container in a pod. The executable specified in the hook command must exist in the target container. The executable can be made available in the container in a couple of ways:
> Can you further elaborate on the "Web schema -> DB schema problem"? I'm not sure what that means.
> 
> —
> Reply to this email directly or view it on GitHub.
>